ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Util/LeakyBucket.hs

```diff
@@ -126,62 +191,77 @@ runAgainstBucket ::
   (MonadDelay m, MonadAsync m, MonadFork m, MonadMask m, NoThunks (m ())) =>
   Config m ->
   (Handlers m -> m a) ->
-  m (State (Config m), a)
+  m (State m, a)
 runAgainstBucket config action = do
+  runThreadVar <- atomically newEmptyTMVar
   bucket <- init config
-  tid <- myThreadId
-  killThreadVar <- newTVarIO Nothing
-  finally
-    ( do
-        startThread killThreadVar bucket tid
-        result <-
-          action $
-            Handlers
-              { fill = (snd <$>) . snapshotFill bucket,
-                setPaused = setPaused bucket,
-                updateConfig = updateConfig killThreadVar bucket tid
-              }
-        state <- snapshot bucket
-        pure (state, result)
-    )
-    (stopThread killThreadVar)
+  bracket
+    (spawnThreadAndMaybeStart runThreadVar bucket)
+    killThread
+    $ \_ -> do
+      result <-
+        action $
+          Handlers
+            { fill = \r t -> (snd <$>) $ snapshotFill bucket r t,
+              setPaused = setPaused bucket,
+              updateConfig = updateConfig runThreadVar bucket
+            }
+      state <- atomicallyWithMonotonicTime $ snapshot bucket
+      pure (state, result)
   where
-    startThread ::
-      StrictTVar m (Maybe (m ())) ->
+    spawnThreadAndMaybeStart ::
+      StrictTMVar m Int ->
       Bucket m ->
-      ThreadId m ->
-      m ()
-    startThread killThreadVar bucket tid =
-      readTVarIO killThreadVar >>= \case
-        Just _ -> error "LeakyBucket: startThread called when a thread is already running"
-        Nothing -> (atomically . writeTVar killThreadVar) =<< leak bucket tid
+      m (Async m ())
+    spawnThreadAndMaybeStart runThreadVar bucket = do
+      tid <- myThreadId
+      atomicallyWithMonotonicTime $ maybeStartThread Nothing runThreadVar bucket
+      async $ leak runThreadVar tid bucket
+
+    killThread :: Async m () -> m ()
+    killThread = uninterruptibleCancel
 
-    stopThread :: StrictTVar m (Maybe (m ())) -> m ()
-    stopThread killThreadVar =
-      readTVarIO killThreadVar >>= \case
-        Just killThread' -> killThread'
-        Nothing -> pure ()
+    -- Start the thread (that is, write to its 'runThreadVar') if it is useful.
+    -- Takes a potential old value of the 'runThreadVar' as first argument,
+    -- which will be increased to help differentiate between restarts.
```

Comment:
From my understanding, the background thread will now always be running, right? So "restart" might be a bit misleading here. Maybe there could be a high-level comment somewhere explaining the lifecycle and state flow of the background thread?

Reply:
This is correct, the thread is now always running. I'm not super satisfies with this solution but I can't think of any solution that fits all our criteria. It is also a thread that mostly does nothing, so I don't think it's such a big deal.

I'm fine getting rid of “start”/“restart” as vocabulary for writing to the TMVar and rewriting to the TMVar, but I don't have any other ideas right now :p Where would you want the comment in question? More details in the description of `leak`?

Reply:
> This is correct, the thread is now always running. I'm not super satisfies with this solution but I can't think of any solution that fits all our criteria. It is also a thread that mostly does nothing, so I don't think it's such a big deal.

I think it is fine, it shouldn't be *that* expensive and it indeed mostly sleeps to occasionally wake up and to trivial arithmetic :+1:

> I'm fine getting rid of “start”/“restart” as vocabulary for writing to the TMVar and rewriting to the TMVar, but I don't have any other ideas right now :p Where would you want the comment in question? More details in the description of `leak`?

It could be a standalone top-level comment, like a [GHC-style note](https://gitlab.haskell.org/ghc/ghc/-/wikis/commentary/coding-style#21-notes-the-basic-idea).

Reply:
Tried in 1d6b842f0dc982e50e318a2cbfef74af7435a2aa. WDYT?

Reply:
Looks great, thanks :+1: